### PR TITLE
Filtering logic

### PR DESCRIPTION
**Pull request by ryliemn**
A good portion of the logic for filtering is here. I figure it's a good time to stop and ask for input on how you would start modularizing things

### CODE REVIEW

File: jQuery/ajax.js [hunk 1]
```diff
@@ -6,6 +6,18 @@ function ajax(citizens) {
...
+    var params = gatherParams();
+    var filteredCitizens = filter(params, event.data.citizens);
+    redrawContent(filteredCitizens);
+  };
+
+  $('#character-select').on('change', { citizens: citizens }, selectHandler);
+  $('#hometown-select').on('change', { citizens: citizens }, selectHandler);
+  $('#size-select').on('change', { citizens: citizens }, selectHandler);
+
+  selectHandler({data: { citizens: citizens }});
```
Comment (ryliemn) [inline jQuery/ajax.js]: I don't really like calling the handler manually like this for initial population of the content

Comment (stffrd) [reply jQuery/ajax.js]: Well. You're not really able to prepopulate it without making an ajax call and manually "redrawing" . A lot of frameworks do this. 

This is just kind of slow because the DOM is the bottleneck. there are only so many times per second you can make changes to the DOM.

File: jQuery/ajax.js [hunk 2]
```diff
@@ -6,6 +6,18 @@ function ajax(citizens) {
...
+
+  var selectHandler = function(event) {
+    var params = gatherParams();
+    var filteredCitizens = filter(params, event.data.citizens);
+    redrawContent(filteredCitizens);
+  };
+
+  $('#character-select').on('change', { citizens: citizens }, selectHandler);
+  $('#hometown-select').on('change', { citizens: citizens }, selectHandler);
+  $('#size-select').on('change', { citizens: citizens }, selectHandler);
```
Comment (stffrd) [inline jQuery/ajax.js]: Don't repeat yourself

`$('#character-select, #hometown-select, #size-select').on('change' ...)`

File: jQuery/ajax.js (hunk 1 above)
Comment (stffrd) [reply jQuery/ajax.js]: Also, when you have multiple listeners like this, you should delegate so instead of having 3 listeners, 1 per element, you have one.  [I've written about this](https://morklympious.github.io/efficiency-event-handling/)

```
<div class="menu">
  <label>
    <select id="character-select" class="menu-select">
  </label>

  . . .
</div>
```

Assuming the other stuff is there, you could actually keep a reference to your top level menu and do some delegation

```
var selectors = {
  menu: ".menu"
}

$(selectors.menu).on('change', { citizens: citizens }, redraw)

```

Comment (stffrd) [reply jQuery/ajax.js]: Where 'redraw' is your handler.

File: jQuery/ajax.js (hunk 2 above)
Comment (ryliemn) [reply jQuery/ajax.js]: Ah nice, didn't know you could do this

File: jQuery/ajax.js [hunk 3]
```diff
@@ -21,6 +33,50 @@ function getUniqueValuesForField(field, citizens) {
 
 function populateSelectWithOptions(select, options) {
   options.forEach(function(option) {
-    $(select).append($("<option>", { value: option, html: option }));
+    $(select).append($('<option>', { value: option, html: option }));
+  });
+}
+
+function gatherParams() {
```
Comment (stffrd) [inline jQuery/ajax.js]: This function is stateful. Which is fine, if you're okay with side effects sometimes. 

Here's how I'd write this.:

``` js
function state(form) {

  // This function is now stateless.
  var character = form.find("#character-select"),
      hometown  = form.find("#hometown-select"),
      size      = form.find("#size-select");

  return [
    { field: "character", value: character.val() },
    { field: "hometown", value: hometown.val() },
    { field: "size", value: size.val() }
  ]
}
```

The function is now "stateless" because it takes a reference to a DOM node via jQuery and uses that node to begin finding the elements of the form it needs. 

So in essence you would pass in the jquery selected menu to get all the fields like so:

``` js
somefunction() {

var values = state($(".menu"))
//or
var whatever = state($(selectors.menu));

}
```

File: jQuery/ajax.js [hunk 4]
```diff
@@ -21,6 +33,50 @@ function getUniqueValuesForField(field, citizens) {
...
+  var hometownValue = $('#hometown-select').val();
+  params.push({field: 'hometown', value: hometownValue});
+
+  var sizeValue = $('#size-select').val();
+  params.push({field: 'size', value: sizeValue});
+
+  return params;
+}
+
+function filter(params, citizens) {
```
Comment (stffrd) [inline jQuery/ajax.js]: Can you just explain to me what this function does in plain english? 

Like, is it looking for a match on a single object in the three-object array we returned with params?

File: jQuery/styles.css [hunk 5]
```diff
@@ -3,7 +3,7 @@ h1 {
   margin-bottom: 20px;
 }
 
-#menu {
+.menu {
```
Comment (stffrd) [inline jQuery/styles.css:6]: I will comment on these when I pull this and build it locally!

File: jQuery/ajax.js [hunk 6]
```diff
@@ -6,6 +6,18 @@ function ajax(citizens) {
   populateSelectWithOptions("#character-select", characters);
   populateSelectWithOptions("#hometown-select", hometowns);
   populateSelectWithOptions("#size-select", sizes);
+
+  var selectHandler = function(event) {
```
Comment (stffrd) [inline jQuery/ajax.js]: Q: Why are you mixing function expressions with declarations? :O

Comment (stffrd) [reply jQuery/ajax.js]: e.g. `var thing = function() {}` vs `thing() {}`

**Review by stffrd — COMMENTED** [review]
There's a lot of refactoring to be done here, and I'm glad this is in "ajax.js" right now. We'll be doing more file splitting soon. 

If you want to, you can split other logic into other files. if you think about it, the AJAX "module" should only really care about pulling down the data and making it available in a meaningful way, you dig?

**Comment by stffrd** [comment]
I remember talking to you about modules, I'm pretty sure. 
[This is important to read (just this chapter)](https://addyosmani.com/resources/essentialjsdesignpatterns/book/#revealingmodulepatternjavascript)

Also read my thing about [namespacing](https://morklympious.github.io/namespacing-javascript/)

because we're going to use these two things to modularize!

File: jQuery/ajax.js (hunk 4 above)
Comment (ryliemn) [reply jQuery/ajax.js]: It's looping through all citizens and returning a list of those that meet the parameters that were gathered by the `gatherParams()` function. It's written in such a  way that if Briana decided to add an extra info field for each citizen (which could definitely happen at some point) the filtering function wouldn't have to change, I'd just have to add a new form element and tell `gatherParams()` how to get the new field.

I'm actually going to do a bit more work on this now, as I'd like to add a freeform text input field so she can search on the names. I'll have to change some things here for that

File: jQuery/ajax.js (hunk 6 above)
Comment (ryliemn) [reply jQuery/ajax.js]: It's because I'm a ding dang dingus

**Comment by ryliemn** [comment]
should I be refactoring all methods that interact with the form fields in the DOM to take the form as a parameter, as you had me do with the `gatherParams()` function?

**Comment by ryliemn** [comment]
Before I start splitting files out, let me know if this attempted modularization makes any sense at all. I thought it made sense to split things out into functions that were doing DOM interactions and functions that were just handling data unaware of the DOM at all, but maybe you had a different idea?

File: jQuery/ajax.js [hunk 7]
```diff
@@ -1,82 +1,113 @@
...
-    var params = gatherParams();
-    var filteredCitizens = filter(params, event.data.citizens);
-    redrawContent(filteredCitizens);
+window.hh = {};
+
+window.hh.selectors = {
+  menu: '.menu',
+  content: '.content',
+  nameInput: '#name-input',
+  characterSelect: '#character-select',
```
Comment (stffrd) [inline jQuery/ajax.js:7]: because I'm an avid hater of camelcase, I would do some shit like this:

``` js
window.hh.selectors = {
  menu: ".menu",
  content: ".content",
  inputs: {
    name: "#name-input"
  },
  selects: {
    character: "#character-select",
    name: "#name-input",
    // etc.
  }
}
```

totally your call.

File: jQuery/ajax.js [hunk 8]
```diff
@@ -1,82 +1,113 @@
...
+  hometownSelect: '#hometown-select',
+  sizeSelect: '#size-select'
+};
+
+/* Defines functions that interact directly with the DOM */
+window.hh.domManipulation = window.hh.domManipulation || (function() {
+  function populateSelectWithOptions(select, options) {
```
Comment (stffrd) [inline jQuery/ajax.js:14]: This function seems needlessly specific. You're taking one element and populating it with other elements. You see where I'm going with this?

consider a world where we have a pure function that accepts two parameters, a parent and a child:

``` js
function _append(parent, child) {
  // Both are wrapped in jQuery objects just in case they were
  // passed in as native DOM elements. 
  $(parent).append($(child));
}
```

And consider that the other thing you're doing is creating new html elements every time based on some string (html or form field values)

``` js
function _create(type, options) {
  return $(type, options); 
}
```

And now you can use this function like so!

(I'm renaming it `populate` because I'm a crochety asshole

``` js
function populate(element, children) {
 children.forEach(function(child) {
    _append( select, _create("<option>", { value: child, html: child } )
  });
}
```

If I wanted to be even more pretentious I might just use a pure function for each lol

``` js
function _each(collection, fn) {
  collection.forEach(fn);
}
```

aaaand boom

``` js
function populate(element, children) {
 _each(children, function(child) {
    _append( select, _create("<option>", { value: child, html: child } )
  });
}
```

Kinda more generic this way. I think I had more fun coding it than actually making it useful. oops.  but at least with this you now have a reusable `_append()` function that can attach new elements to other elements, you know?

**Review by stffrd — COMMENTED** [review]
still got more to say, just some quick and dirty and mostly useless feedback loloolololololol

Also I think you're on the right track with "Stuff that touches the dom in one module and stuff that does data processing in another"

File: jQuery/ajax.js [hunk 9]
```diff
@@ -1,82 +1,113 @@
...
-
-  populateSelectWithOptions("#character-select", characters);
-  populateSelectWithOptions("#hometown-select", hometowns);
-  populateSelectWithOptions("#size-select", sizes);
-
-  var selectHandler = function(event) {
-    var params = gatherParams();
-    var filteredCitizens = filter(params, event.data.citizens);
-    redrawContent(filteredCitizens);
+window.hh = {};
```
Comment (stffrd) [inline jQuery/ajax.js]: let's call this bad boy `window.honey`. I hate hh. 

I hate it. 

plus everything looks BETTER WITH IT
`window.honey.ajax`
`window.honey.logic`
`window.honey.dom`

File: jQuery/ajax.js (hunk 7 above)
Comment (stffrd) [reply jQuery/ajax.js:7]: Also, if these selectors are only being used in a single function, it's an acceptable practice to put them as a variable at the top of the module where they make an appearance.

File: jQuery/ajax.js (hunk 9 above)
Comment (ryliemn) [reply jQuery/ajax.js]: OKAY

File: jQuery/ajax.js [hunk 10]
```diff
@@ -1,82 +1,113 @@
...
-  var filteredCitizens = citizens.filter(function(citizen) {
-    var matchesFilter = true;
-
-    for (var i = 0; i < params.length && matchesFilter; i++) {
-      var param = params[i];
-
-      var field = param['field'];
-      var citizenValue = citizen[field];
-      var paramValue = param['value'];
+window.hh.ajax = window.hh.ajax || function(citizens) {
```
Comment (stffrd) [inline jQuery/ajax.js]: Is this a module? you're not returning anything at the end of it. No API here? I also don't see you immediately calling it. 

`window.hh.ajax = window.hh.ajax || function(citizens) {` should be 
`window.hh.ajax = window.hh.ajax || (function(citizens) {`

Comment (stffrd) [reply jQuery/ajax.js]: with the closing paren and instant execution at the end e.g. `})()`

Comment (ryliemn) [reply jQuery/ajax.js]: I figured since the module is literally just going to expose that one function to be called in the ajax callback, it'd be fine to make this refer just to the function itself. I dunno

File: jQuery/ajax.js [hunk 11]
```diff
@@ -1,82 +1,113 @@
...
+  var selectors = window.hh.selectors;
+  var DOM = window.hh.domManipulation;
```
Comment (stffrd) [inline jQuery/ajax.js]: So this is where we get into _dependency injection_ with modules. You can [read this](https://morklympious.github.io/dependency-injection-what/) if you want 

it's a bad idea to declare this explicitly in the module. Instead, we want these module patterns to work for us, so the idea is to pass in the thing you're going to use. For example, this is a very common pattern for a module using jQuery:

``` js
window.utilities = window.utilities || (function($) {

...

})(jQuery)
```

Notice how I passed in the global reference to jQuery so it populates the parameter I defined as `$`? You'll want to do the same with your `ajax` module. 

e.g.

``` js
(function(citizens, logic) {

// I can now use window.hh.logic's API, available through the logic param
// No declaration needed. 

}( /*whatever citizens is */, window.hh.logic));
```

**Review by stffrd — COMMENTED** [review]
We're slowly beating your code into shape my friend.

File: jQuery/ajax.js (hunk 11 above)
Comment (ryliemn) [reply jQuery/ajax.js]: ah yeah, I remember this now.

File: jQuery/ajax.js (hunk 8 above)
Comment (ryliemn) [reply jQuery/ajax.js:14]: As we continue onwards, if a similar pattern pops up anywhere else I'll do the generalization you've suggested here. I'd still like to have the very specific function there (`populateSelectWithOptions`), just delegating to the more generalized ones, for readability purposes

File: jQuery/ajax.js (hunk 11 above)
Comment (ryliemn) [inline jQuery/ajax.js]: You'll notice in ```main.js``` I'm just passing ```window.honey.ajax``` directly into the callback

File: jQuery/ajax.js (hunk 10 above)
Comment (ryliemn) [reply jQuery/ajax.js]: You'll notice in `main.js` I'm just passing `window.honey.ajax` directly into the callback

Comment (ryliemn) [reply jQuery/ajax.js]: ah, I see, can't really do dependency injection as cleanly this way. I'll refactor

**Comment by ryliemn** [comment]
more updates. Possibly good to move on?
